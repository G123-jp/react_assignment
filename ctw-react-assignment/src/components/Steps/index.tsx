import classnames from 'classnames'
import { ReactNode } from 'react'

import styles from './index.module.less'

interface StepOptionItem {
  title: string
  content: ReactNode
}

interface Props {
  current: number
  options: StepOptionItem[]
}

const Steps = ({ current = 0, options = [] }: Props): JSX.Element => {
  return (
    <div className={styles.steps}>
      {options.map((s, index) => {
        return (
          <div
            className={classnames(styles.steps__item, {
              [styles.steps__item_active]: current === index
            })}
            key={index}
          >
            {s.title}
          </div>
        )
      })}
    </div>
  )
}

export default Steps
