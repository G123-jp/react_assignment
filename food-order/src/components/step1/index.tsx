import React, { useCallback, useContext, useMemo, useRef, useState } from 'react';
import type { SyntheticEvent, ChangeEventHandler } from 'react';
import { OrderContext } from '../../store';
import './step1.css';
import breakfast from '../../assets/dumpling.png';
import lunch from '../../assets/curry.png';
import dinner from '../../assets/salad.png';

export const OrderStep1 = () => {
    const [currentSelectedMeal, setCurrentSelectedMeal] = useState('breakfast');
    const orderContext = useContext(OrderContext);
    const numOfPeopleRef = useRef('1');
    const mealClassNames = useMemo(() => {
        const className = 'page1-flex page1-box-wrapper page1-box-item';
        const selectClassName = ' page1-box-item-selected';
        const firstClassName = className + (currentSelectedMeal === 'breakfast' ? selectClassName : '');
        const secondClassName = className + (currentSelectedMeal === 'lunch' ? selectClassName : '');
        const thirdClassName = className + (currentSelectedMeal === 'dinner' ? selectClassName : '');
        return [firstClassName, secondClassName, thirdClassName];
    }, [currentSelectedMeal]);
    const onClickMeal = useCallback((e: SyntheticEvent, index: string) => {
        console.log('event target = ', e + 'index ', index);
        setCurrentSelectedMeal(index);
    }, []);

    const onNuOfPeopleChange = useCallback<ChangeEventHandler<HTMLSelectElement>>((value) => {
        numOfPeopleRef.current = value.target.value;
    }, []);

    const onClickNext = useCallback(() => {
        orderContext && orderContext.onClickNext();
    }, []);

    return (
        <div className="page page1-element">
            <h1 className="page1-heading-title">ORDER FOOD</h1>
            <h3 className="page1-hint-title">Please select a meal</h3>
            <div className="page1-flex page1-row-flex-center page1-element-container">
                <div className={mealClassNames[0]} onClick={(e) => onClickMeal(e, 'breakfast')}>
                    <img src={breakfast}></img>
                    <p>Breakfast</p>
                </div>
                <div className={mealClassNames[1]} onClick={(e) => onClickMeal(e, 'lunch')}>
                    <img src={lunch}></img>
                    <p>Lunch</p>
                </div>
                <div className={mealClassNames[2]} onClick={(e) => onClickMeal(e, 'dinner')}>
                    <img src={dinner}></img>
                    <p>Dinner</p>
                </div>
            </div>
            <h3 className="page1-hint-title">Please enter number of people</h3>
            <select
                className="select"
                required={true}
                aria-required="true"
                defaultValue={numOfPeopleRef.current}
                onChange={onNuOfPeopleChange}
            >
                <option value="1">1</option>
                <option value="2">2</option>
                <option value="3">3</option>
                <option value="4">4</option>
                <option value="5">5</option>
                <option value="6">6</option>
                <option value="7">7</option>
                <option value="8">8</option>
                <option value="9">9</option>
                <option value="10">10</option>
            </select>
            <div className="page1-button-normal page1-flex page1-right-button" onClick={onClickNext}>
                Next
            </div>
        </div>
    );
};
