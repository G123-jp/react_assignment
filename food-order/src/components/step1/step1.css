.page {
    text-align: center;
    flex-direction: column;
    align-items: center;
    justify-content: flex-start;
    display: flex;
    height: 100%;
    width: 100%;
}

.page1-hint-title {
  font-size: 20px;
  font-weight: 600;
  font-family: "Poppins, Sans-serif";
  color: #333;
  margin-top: 40px;
}

.page1-row-flex-center {
  flex-direction: row;
  justify-content: center;
  align-items: center;
}
.page1-row-flex-between {
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.page1-heading-title {
    box-sizing: border-box;
    padding: 0;
    margin-top: 60px;
    font-family: 'Cinzel', Sans-serif;
    font-size: 75px;
    font-weight: 600;
    text-transform: capitalize;
    letter-spacing: 1.5px;
    text-shadow: -1px 9px 11px rgb(0 0 0 / 11%);
    color: #333;
    line-height: 1;
    text-size-adjust: 100%;
}

.page1-element {
  background-color: transparent;
  background-image: linear-gradient(180deg, #F2F2F4 0%, #F7F6F2 100%);
}

.page1-element-populated {
  border-style: solid;
  border-width: 2px 2px 2px 2px;
  /* border-color: #ddccdd26; */
  box-shadow: 0px 0px 10px 0px rgb(0 0 0 / 9%);
  transition: background 0.3s, border 0.3s, border-radius 0.3s, box-shadow 0.3s;
  margin: 10px 10px 10px 10px;
  padding: 0px 0px 17px 0px;
}

.page1-element-container {
  width: 77%;
  height: 280px;
}

.page1-option {
  font-weight: normal;
  display: block;
  white-space: nowrap;
  min-height: 1.2em;
  padding: 0px 2px 1px;
  color: #878787;
}

.page1-select {
  width: 250px;
  background-color: #f1f1f5;
  background-image: url("data:image/svg+xml,%3Csvg xmlns=%22http://www.w3.org/2000/svg%22 fill=%22%23212529%22 viewBox=%220 0 1e3 1e3%22%3E%3Cpath d=%22M5e2 10 227.8 391.1h544.4L5e2 10zm0 980 272.2-381.1H227.8L5e2 990z%22/%3E%3C/svg%3E");
  background-repeat: no-repeat;
  background-position: 100% 15px;
  background-size: 45px 14px;
  padding-right: 60px;
  color: #878787;
  padding: 0 15px;
  height: 45px;
  appearance: none;
  -moz-appearance: none;
  -webkit-appearance: none;
}
.page1-flex {
  display: flex;
}

.page1-button-normal {
    background-color: #437f57;
    font-family: "Poppins, Sans-serif";
    font-size: 16px;
    font-weight: 500;
    text-transform: capitalize;
    line-height: 24px;
    letter-spacing: 0.7px;
    word-spacing: 0;
    border-radius: 20px;
    width: 150px;
    height: 50px;
    text-align: center;
    justify-content: center;
    align-items: center;
    fill: #f2f2f4; 
    color: #f2f2f4;
    transition: 0.3s;
}

.page1-button-normal:hover {
  background-color: #FABB3D;
  width: 135px;
  height: 45px;
}

.page1-right-button {
  margin-top: 100px;
}

.page1-box-wrapper {
  width: 20%;
  height: 100%;
  align-items: center;
  flex-direction: column;
  justify-content: flex-end;
  margin-right: 20px;
}

.page1-box-item {
  text-align: center;
  border-radius: 20px;
  transition: 0.3s;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
}

.page1-box-item-selected {
  text-align: center;
  border-width: 2px;
  border-style: solid;
  border-color: #437f57;
  background-color: #437f57;
}


.page1-box-item img {
  width: 159px;
  height: 159px;
  transform: scale(1);
  transition: all .4s ease;
}
.page1-box-item p {
  margin-top: 20px;
  font-size: 25px;
  font-weight: 300;
  font-family: "Poppins, Sans-serif";
  color: #444;
  transition: all .4s ease;
}

.page1-box-item-selected img {
  transform: scale(1.2);
}

.page1-box-item-selected p {
  transform: translateY(-20px);
  color: #F1FFF4;
}

@keyframes fadeInUp {
    from {
        opacity: 0;
        transform: translate3d(0, 100%, 0);
    }
    to {
        opacity: 1;
        transform: none;
    }
}
.fadeInUp {
    animation-name: fadeInUp;
}
