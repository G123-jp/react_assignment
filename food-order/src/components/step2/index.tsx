import React, { useCallback, useContext, useRef } from 'react';
import type { ChangeEventHandler } from 'react';
import { OrderContext } from '../../store';

import './step2.css';
import leftsalad from '../../assets/leftsalad.png';

interface IOrderStepProps {
    restaurants: string[];
}

export const OrderStep2 = (props: IOrderStepProps) => {
    const { restaurants } = props;
    const orderContext = useContext(OrderContext);
    const selectRestaurantRef = useRef(restaurants[0]);
    const onClickPrevious = useCallback(() => {
        orderContext && orderContext.onClickPrevious();
    }, []);

    const onNuOfPeopleChange = useCallback<ChangeEventHandler<HTMLSelectElement>>((value) => {
        selectRestaurantRef.current = value.target.value;
    }, []);

    const onClickNext = useCallback(() => {
        orderContext && orderContext.onClickNext();
    }, []);

    return (
        <div className="page page2-flex page2-row-flex-center">
            <div className="page2-flex page2-col-flex-center page2-col-30">
                <h1 className="page2-heading-title">Online Reservation</h1>
                <img src={leftsalad} width={620} height={543} alt="" />
            </div>
            <div className="page2-flex page2-col-start-start page2-col-30 page2-col-right">
                <div className="page2-flex page2-col-start-start">
                    <label className="page2-right-form-label">
                        Please select a meal<small className="page2-right-form-required">*</small>
                    </label>
                    <select
                        className="select"
                        required={true}
                        aria-required="true"
                        defaultValue={selectRestaurantRef.current}
                        onChange={onNuOfPeopleChange}
                    >
                        {restaurants.map((value: string, index: number) => (
                            <option key={index} value={value}>
                                {value}
                            </option>
                        ))}
                    </select>
                </div>
                <div className="page2-flex page2-right-form-button-box">
                    <div className="page2-flex page2-button-normal" onClick={onClickNext}>
                        Next
                    </div>
                    <div className="page2-flex page2-right-form-button-cancel" onClick={onClickPrevious}>
                        Previous
                    </div>
                </div>
            </div>
        </div>
    );
};
