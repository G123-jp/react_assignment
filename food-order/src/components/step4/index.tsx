import React, { useCallback, useContext, useRef } from 'react';
import { OrderContext } from '../../store';
import { OrderedDishItem } from '../../store/model';

import './step4.css';

interface IOrderStep4Props {
    orderDishes: OrderedDishItem[];
    onCloseCallback: () => void;
}

export const OrderStep4 = (props: IOrderStep4Props) => {
    const { orderDishes, onCloseCallback } = props;
    const orderContext = useContext(OrderContext);
    const buttonRef = useRef<HTMLButtonElement>(null);

    const onCloseClick = useCallback(() => {
        onCloseCallback();
    }, [onCloseCallback]);

    const onClickButton = useCallback(() => {
        const button = buttonRef.current;
        button?.classList.remove('animate');

        button?.classList.add('animate');
        button?.classList.add('animate');
        setTimeout(function () {
            button?.classList.remove('animate');
        }, 6000);
    }, []);

    return (
        <div className="page4-backgroud">
            <div className="page4">
                <div className="page4-left-box">
                    <img
                        src={require('../../assets/' + orderDishes[0].imageIndex + '.png')}
                        className="page4-left-dish-img"
                    />
                </div>
                <div className="page4-order-box">
                    <div className="page4-close-box" onClick={onCloseClick}>
                        {' '}
                        <i className="fa-solid fa-xmark"></i>
                    </div>
                    <h1 className="page4-heading-title">Please check your order</h1>
                    <ul className="page4-form">
                        <li className="page4-form-li page4-form-li-row">
                            <p className="page4-form-li-row-title">Meal</p>
                            <p className="page4-form-li-row-content">{orderContext?.contextMealType}</p>
                        </li>
                        <li className="page4-form-li page4-form-li-row">
                            <p className="page4-form-li-row-title">No. of. people</p>
                            <p className="page4-form-li-row-content">{orderContext?.contextNumOfPeople}</p>
                        </li>
                        <li className="page4-form-li page4-form-li-row">
                            <p className="page4-form-li-row-title">Restaurant</p>
                            <p className="page4-form-li-row-content">{orderContext?.contextRestaurant}</p>
                        </li>
                    </ul>
                    <div className="page4-dishes-box">
                        <p className="page4-form-li-row-title">Dishes</p>
                        <div className="page4-dish-preview-box">
                            {orderDishes.map((value) => {
                                return (
                                    <div key={value.id} className="page4-form-li page4-dish-preview-item">
                                        <p className="page4-dish-preview-title">{value.name}</p>
                                        <p className="page4-dish-preview-subtitle">{`x ${value.orderNum}`}</p>
                                    </div>
                                );
                            })}
                        </div>
                    </div>
                    <button ref={buttonRef} onClick={onClickButton} className="page4-button-normal success">
                        Order Now
                    </button>
                </div>
            </div>
        </div>
    );
};
