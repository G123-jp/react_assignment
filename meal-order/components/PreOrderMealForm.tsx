import { DishList } from "@/pages/api/dishes";
import { RestaurantList } from "@/pages/api/restaurants";
import { Dish, MealType, SelectedDishes } from "@/shared/types";
import { ReactNode, useEffect, useReducer, useState } from "react";
import OrderSummary from "./steps/OrderSummary";
import Step1Form from "./steps/Step1Form";
import Step2Form from "./steps/Step2Form";
import Step3Form from "./steps/Step3Form";

const FormProgressItem = ({
  isHighlighted,
  children,
}: {
  isHighlighted: boolean;
  children: ReactNode;
}) => {
  if (isHighlighted) {
    return (
      <span className="mx-1 bg-slate-500 text-white font-bold p-2">
        {children}
      </span>
    );
  } else {
    return (
      <span className="mx-1 bg-slate-200 text-slate-500 p-2">{children}</span>
    );
  }
};

const FormProgress = ({ currentStep }: { currentStep: number }) => {
  return (
    <div className="flex flex-row justify-center mt-4">
      <FormProgressItem isHighlighted={currentStep === 0}>
        Step 1
      </FormProgressItem>
      <span className="py-2 px-1">{">"}</span>
      <FormProgressItem isHighlighted={currentStep === 1}>
        Step 2
      </FormProgressItem>
      <span className="py-2 px-1">{">"}</span>
      <FormProgressItem isHighlighted={currentStep === 2}>
        Step 3
      </FormProgressItem>
      <span className="py-2 px-1">{">"}</span>
      <FormProgressItem isHighlighted={currentStep === 3}>
        Review
      </FormProgressItem>
    </div>
  );
};

const NavigationButtons = ({
  onNext,
  onPrev,
  hidePrev = false,
  showSubmit = false,
  prevText = "Previous",
  nextText = "Next",
}: {
  onNext: () => void;
  onPrev: () => void;
  hidePrev?: boolean;
  showSubmit?: boolean;
  prevText?: string;
  nextText?: string;
}) => {
  return (
    <div className="flex flex-row justify-between mt-4">
      <button
        onClick={onPrev}
        className={`bg-blue-400 text-white p-2 rounded-xl font-bold ${
          hidePrev ? "invisible" : ""
        }`}
      >
        {prevText}
      </button>
      <button
        onClick={onNext}
        className={"bg-blue-400 text-white p-2 rounded-xl font-bold"}
      >
        {nextText}
      </button>
    </div>
  );
};

const NUMBER_OF_STEPS = 4;

interface StateType {
  selectedMealType: MealType;
  numOfPeople: number;
  selectedRestaurant: string;
  selectedDishes: SelectedDishes; // dish id mapping to Dish Serving
}

const initialState: StateType = {
  selectedMealType: "breakfast",
  numOfPeople: 1,
  selectedRestaurant: "",
  selectedDishes: {},
};

type ActionType =
  | { type: "select_meal_type"; payload: { mealType: MealType } }
  | { type: "select_num_people"; payload: { numOfPeople: number } }
  | { type: "select_restaurant"; payload: { restaurant: string } }
  | {
      type: "update_dish_order";
      payload: { dish: Dish; numberOfServing: number };
    };

const reducer = (state: StateType, action: ActionType): StateType => {
  if (action.type === "select_meal_type") {
    return {
      ...state,
      selectedMealType: action.payload.mealType,
      selectedRestaurant: "", // reset restaurant selection
      selectedDishes: {}, // reset dish selection
    };
  } else if (action.type === "select_num_people") {
    return {
      ...state,
      numOfPeople: action.payload.numOfPeople,
    };
  } else if (action.type === "select_restaurant") {
    return {
      ...state,
      selectedRestaurant: action.payload.restaurant,
      selectedDishes: {}, // reset dish selection
    };
  } else if (action.type === "update_dish_order") {
    const {
      payload: { dish, numberOfServing },
    } = action;

    const newSelectedDishes = { ...state.selectedDishes };
    if (numberOfServing <= 0) {
      delete newSelectedDishes[`${action.payload.dish.id}`];
    } else {
      newSelectedDishes[`${action.payload.dish.id}`] = {
        dish,
        numberOfServing,
      };
    }
    return {
      ...state,
      selectedDishes: newSelectedDishes,
    };
  }
  throw Error("unknown action");
};

const useRestaurants = (mealType: MealType) => {
  const [data, setData] = useState<RestaurantList | null>(null);
  const [isLoading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    setLoading(true);
    fetch(`/api/restaurants?mealType=${mealType}`)
      .then((res) => res.json())
      .then((data) => {
        setData(data);
        setLoading(false);
      });
  }, [mealType]);

  return {
    restaurants: data ? data.restaurants : [],
    isLoading,
  };
};

const useDishes = (mealType: MealType, restaurant: string) => {
  const [data, setData] = useState<DishList | null>(null);
  const [isLoading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    setLoading(true);
    fetch(`/api/dishes?mealType=${mealType}&restaurant=${restaurant}`)
      .then((res) => res.json())
      .then((data) => {
        setData(data);
        setLoading(false);
      });
  }, [mealType, restaurant]);

  return {
    dishes: data ? data.dishes : [],
    isLoading,
  };
};

export default function PreOrderMealForm() {
  const [currentStep, setCurrentStep] = useState<number>(0);
  const [state, dispatch] = useReducer(reducer, initialState);
  const { selectedMealType, numOfPeople, selectedRestaurant, selectedDishes } =
    state;
  const { restaurants } = useRestaurants(selectedMealType);
  const { dishes } = useDishes(selectedMealType, selectedRestaurant);

  const CurrentForm = () => {
    switch (currentStep) {
      case 0:
        return (
          <Step1Form
            selectedMealType={selectedMealType}
            onMealTypeSelected={(mealType) =>
              dispatch({ type: "select_meal_type", payload: { mealType } })
            }
            numOfPeople={numOfPeople}
            onNumOfPeopleChanged={(numOfPeople) => {
              dispatch({ type: "select_num_people", payload: { numOfPeople } });
            }}
          />
        );
      case 1:
        return (
          <Step2Form
            onRestaurantSelected={(restaurant) => {
              dispatch({ type: "select_restaurant", payload: { restaurant } });
            }}
            selectedRestaurant={selectedRestaurant}
            restaurants={restaurants}
          />
        );
      case 2:
        return (
          <Step3Form
            availableDishes={dishes}
            selectedDishes={selectedDishes}
            onUpdateDish={(dish, numberOfServing) => {
              dispatch({
                type: "update_dish_order",
                payload: { dish, numberOfServing },
              });
            }}
          />
        );
      case 3:
      default:
        return <OrderSummary />;
    }
  };

  const goToNextStep = () => {
    if (currentStep < NUMBER_OF_STEPS - 1) {
      setCurrentStep(currentStep + 1);
    }
  };

  const goToPrevStep = () => {
    if (currentStep > 0) {
      setCurrentStep(currentStep - 1);
    }
  };

  return (
    <>
      <h1 className="text-2xl">Pre-order Your Meal</h1>
      <FormProgress currentStep={currentStep} />
      <CurrentForm />
      <NavigationButtons
        onNext={goToNextStep}
        onPrev={goToPrevStep}
        hidePrev={currentStep === 0}
        nextText={currentStep === NUMBER_OF_STEPS - 1 ? "Submit" : undefined}
      />
    </>
  );
}
