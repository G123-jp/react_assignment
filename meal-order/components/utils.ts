import { Dish, MealType, SelectedDishes } from "@/common/types";
import { countTotalNumberOfServings } from "@/common/utils";

export enum STEPS {
    Step1,
    Step2,
    Step3,
    Review,
    SubmitOrder,
}
  
type FormValidity = {
    isFormValid: boolean;
    errorMessage?: string;
};

export interface StateType {
    selectedMealType: MealType | null;
    numOfPeople: number;
    selectedRestaurant: string;
    selectedDishes: SelectedDishes; // dish id mapping to Dish Serving
    currentStep: STEPS;
    formValidity: FormValidity;
}

export const initialState: StateType = {
    selectedMealType: null,
    numOfPeople: 1,
    selectedRestaurant: "",
    selectedDishes: {},
    currentStep: STEPS.Step1,
    formValidity: {
        isFormValid: false,
    },
};

type ActionType =
    | { type: "go_to_next_step" }
    | { type: "go_to_prev_step" }
    | { type: "select_meal_type"; payload: { mealType: MealType } }
    | { type: "select_num_people"; payload: { numOfPeople: number } }
    | { type: "select_restaurant"; payload: { restaurant: string } }
    | {
        type: "update_dish_order";
        payload: { dish: Dish; numberOfServing: number };
    };

const getValidityState = (
    state: StateType,
    withErrorMessage: boolean = true // when we want to show/hide error message for UX purpose (e.g. when changing step, we should hide the error message)
): FormValidity => {
    const {
        selectedMealType,
        numOfPeople,
        selectedRestaurant,
        selectedDishes,
        currentStep,
    } = state;
    let isFormValid = true;
    let errorMessage: string | undefined;
    if (currentStep === STEPS.Step1) {
        if (!selectedMealType) {
        isFormValid = false;
        if (withErrorMessage) {
            errorMessage = "A meal type has to be selected";
        }
        }
        if (numOfPeople < 1 || numOfPeople > 10) {
        isFormValid = false;
        if (withErrorMessage) {
            errorMessage = "Number of people has to be between 1 to 10";
        }
        }
    } else if (currentStep === STEPS.Step2) {
        if (!selectedRestaurant) {
        isFormValid = false;
        if (withErrorMessage) {
            errorMessage = "You have to select a restaurant";
        }
        }
    } else if (currentStep === STEPS.Step3) {
        if (countTotalNumberOfServings(selectedDishes) < numOfPeople) {
        isFormValid = false;
        if (withErrorMessage) {
            if (numOfPeople === 1) {
            errorMessage = "You need to order at least 1 serving";
            } else {
            errorMessage = `You need to order at least ${numOfPeople} of servings for ${numOfPeople} people.`;
            }
        }
        }
    }
    
    return { isFormValid, errorMessage };
};
    
export const reducer = (state: StateType, action: ActionType): StateType => {
    const newState = { ...state };
    if (action.type === "go_to_prev_step") {
        if (state.currentStep > STEPS.Step1) {
        newState.currentStep = state.currentStep - 1;
        }
    } else if (action.type === "go_to_next_step") {
        if (state.currentStep <= STEPS.Review) {
        newState.currentStep = state.currentStep + 1;
        }
    } else if (action.type === "select_meal_type") {
        newState.selectedMealType = action.payload.mealType;
        newState.selectedRestaurant = ""; // reset restaurant selection
        newState.selectedDishes = {}; // reset dish selection
    } else if (action.type === "select_num_people") {
        newState.numOfPeople = action.payload.numOfPeople;
    } else if (action.type === "select_restaurant") {
        newState.selectedRestaurant = action.payload.restaurant;
        newState.selectedDishes = {}; // reset dish selection
    } else if (action.type === "update_dish_order") {
        const {
        payload: { dish, numberOfServing },
        } = action;
    
        const newSelectedDishes = { ...state.selectedDishes };
        if (numberOfServing <= 0) {
        delete newSelectedDishes[`${action.payload.dish.id}`];
        } else {
        newSelectedDishes[`${action.payload.dish.id}`] = {
            dish,
            numberOfServing,
        };
        }
    
        newState.selectedDishes = newSelectedDishes;
    } else {
        throw Error("unknown action");
    }
    const withErrorMessage =
        action.type !== "go_to_next_step" && action.type !== "go_to_prev_step"; // hide error message when step has just changed
    newState.formValidity = getValidityState(newState, withErrorMessage);
    return newState;
};
