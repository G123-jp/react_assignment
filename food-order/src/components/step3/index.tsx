import React, { useCallback, useContext, useEffect, useMemo, useState } from 'react';
import type { SyntheticEvent } from 'react';
import { OrderContext } from '../../store';
import { DishItem, OrderedDishItem } from '../../store/model';
import Modal from 'react-modal';
import { OrderStep4 } from '../step4';

import './step3.css';
import dishJson from '../../assets/dishes.json';

Modal.setAppElement('#root');

interface IOrderStep3Props {
    dishes: DishItem[];
}

interface IOrderDishesNum {
    [index: string]: OrderedDishItem;
}

const caluImageIndex = (indexNumber: number) => {
    const imageNameIndex = (indexNumber % 8) + 1;
    return imageNameIndex;
};

export const OrderStep3 = (props: IOrderStep3Props) => {
    const { dishes } = props;
    const dataSource = dishJson.dishes;
    const [modalOpenState, setModalOpenState] = useState(false);
    const [orderNumState, setOrderNumState] = useState(0);

    const orderContext = useContext(OrderContext);
    const contextNumOfPeople = orderContext?.contextNumOfPeople || 0;
    const [showValidMessageState, setShowValidMessageState] = useState(false);
    const [showingAlertState, setShowingAlertState] = useState(false);
    const [initialOrderDishesState] = useState(() => {
        const initialOrderDishes: IOrderDishesNum = {};
        dataSource.forEach((value) => {
            const key = value.id;
            const imageNameIndex = caluImageIndex(key);
            initialOrderDishes[key] = { orderNum: 0, id: value.id, name: value.name, imageIndex: imageNameIndex };
        });
        return initialOrderDishes;
    });
    const [orderDishesState, setOrderDishesState] = useState<IOrderDishesNum>(initialOrderDishesState);

    const orderedDishes = useMemo(() => {
        const dishesArray = Object.values<OrderedDishItem>(orderDishesState);
        return dishesArray.filter((element) => {
            return element.orderNum > 0;
        });
    }, [orderDishesState]);
    console.log('orderedDishes !!!! : ', orderedDishes);

    const onClickUpdateNum = useCallback(
        (e: SyntheticEvent, index: number, isPlus: boolean) => {
            const orderDishCopy = { ...orderDishesState };
            const orderingDish = orderDishCopy[index];
            const newNum = orderingDish.orderNum + (isPlus ? 1 : -1);
            orderDishCopy[index] = { ...orderingDish, ...{ orderNum: Math.max(newNum, 0) } };
            setOrderNumState((pre) => {
                return pre + (isPlus ? 1 : -1);
            });
            setOrderDishesState(orderDishCopy);
        },
        [orderDishesState]
    );

    const onClickPrevious = useCallback(() => {
        orderContext?.onClickPrevious();
        setTimeout(() => {
            setOrderDishesState(initialOrderDishesState);
        }, 500);
    }, []);

    const onClickNext = useCallback(() => {
        console.log('currentDishNum :', orderNumState + ' contextNumOfPeople :', contextNumOfPeople);
        if (orderNumState < contextNumOfPeople) {
            if (!showingAlertState) {
                setShowingAlertState(true);
            }
            setShowValidMessageState(true);
            return;
        }
        setShowValidMessageState(false);
        setModalOpenState(true);
    }, [contextNumOfPeople, orderNumState, showingAlertState]);
    const invalidMessage = useMemo(() => {
        return ` Please order at least ${contextNumOfPeople} meals!`;
    }, [contextNumOfPeople]);

    const alertClassName = useMemo(() => {
        return `page3-next-invalid${showValidMessageState ? ' fadeInOut' : ''}`;
    }, [showValidMessageState]);

    const onRequestCloseCallback = useCallback(() => {
        setModalOpenState(false);
    }, []);

    useEffect(() => {
        const timerId = setTimeout(() => {
            setShowingAlertState(false);
            setShowValidMessageState(false);
        }, 5000);
        return () => {
            clearTimeout(timerId);
        };
    }, [showingAlertState]);
    console.log('root setModalOpenState : ', modalOpenState);
    return (
        <div className="page3-backgroud">
            <div className="page3 page3-flex page3-col-flex-center">
                <h1 className="page3-heading-title">Please select a dish</h1>
                <label className="page3-hint-title">食べ物を無駄にしない</label>
                <div className="page3-flex page3-row-flex-center page3-element-container">
                    {dishes.map((value) => {
                        const key = value.id;
                        const imageNameIndex = caluImageIndex(key);
                        const orderNum = orderDishesState[key].orderNum;
                        return (
                            <div key={key} className="page3-col-25">
                                <img
                                    src={require('../../assets/' + imageNameIndex + '.png')}
                                    className="page3-col-25-inner-img"
                                />
                                <div className="page3-flex page3-col-flex-end page3-col-25-inner">
                                    <p className="page3-dish-item-title">{value.name}</p>
                                    <p className="page3-dish-item-subtitle">Mus Natoque Quisque Tincidunt</p>
                                    <div className="page3-flex page3-row-flex-between page3-dish-item-num-box">
                                        <label className="page3-dish-item-num">{`${orderNum}`}</label>
                                        <div className="page3-flex page3-row-flex-between page3-icon-button-box">
                                            <div
                                                className="page3-flex page3-icon-minus-button"
                                                onClick={(e) => {
                                                    onClickUpdateNum(e, value.id, false);
                                                }}
                                            >
                                                {' '}
                                                <i className="fa fa-minus"></i>
                                            </div>
                                            <div
                                                className="page3-flex page3-icon-button-normal"
                                                onClick={(e) => {
                                                    onClickUpdateNum(e, value.id, true);
                                                }}
                                            >
                                                {' '}
                                                <i className="fa fa-plus"></i>
                                            </div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        );
                    })}
                </div>
            </div>
            <div className="page3-button-next" onClick={onClickNext}>
                Next
            </div>
            <div className="page3-button-cancel" onClick={onClickPrevious}>
                Previous
            </div>
            <div className={alertClassName}>
                <i className="fa-solid fa-circle-exclamation"></i>
                {invalidMessage}
            </div>
            <Modal
                isOpen={modalOpenState}
                onRequestClose={onRequestCloseCallback}
                className="page3-modal-content"
                overlayClassName="page3-modal-overlay"
            >
                <OrderStep4 orderDishes={orderedDishes} onCloseCallback={onRequestCloseCallback}></OrderStep4>
            </Modal>
        </div>
    );
};
