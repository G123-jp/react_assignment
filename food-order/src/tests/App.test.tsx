import React from 'react';
import { render, screen } from '@testing-library/react';
import { OrderStep1 } from '../components/step1';
import { create } from 'react-test-renderer';

test('renders first page', () => {
    const step1Components = create(<OrderStep1 animateName="" />).toJSON();
    expect(step1Components).toMatchSnapshot();
});
