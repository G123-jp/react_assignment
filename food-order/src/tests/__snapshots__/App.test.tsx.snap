// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders first page 1`] = `
<div
  className="page1-backgroud"
>
  <div
    className="page1"
  >
    <h1
      className="page1-heading-title"
    >
      Order Food
    </h1>
    <label
      className="page1-hint-title"
    >
      Please select a meal
    </label>
    <div
      className="page1-flex page1-row-flex-center page1-element-container"
    >
      <div
        className="page1-flex page1-box-wrapper page1-box-item page1-box-item-selected"
        onClick={[Function]}
      >
        <img
          src="dumpling.png"
        />
        <p>
          Breakfast
        </p>
      </div>
      <div
        className="page1-flex page1-box-wrapper page1-box-item"
        onClick={[Function]}
      >
        <img
          src="curry.png"
        />
        <p>
          Lunch
        </p>
      </div>
      <div
        className="page1-flex page1-box-wrapper page1-box-item"
        onClick={[Function]}
      >
        <img
          src="salad.png"
        />
        <p>
          Dinner
        </p>
      </div>
    </div>
    <h3
      className="page1-hint-title"
    >
      Please enter number of people
    </h3>
    <select
      aria-required="true"
      className="page1-select"
      defaultValue="1"
      onChange={[Function]}
      required={true}
    >
      <option
        value="1"
      >
        1
      </option>
      <option
        value="2"
      >
        2
      </option>
      <option
        value="3"
      >
        3
      </option>
      <option
        value="4"
      >
        4
      </option>
      <option
        value="5"
      >
        5
      </option>
      <option
        value="6"
      >
        6
      </option>
      <option
        value="7"
      >
        7
      </option>
      <option
        value="8"
      >
        8
      </option>
      <option
        value="9"
      >
        9
      </option>
      <option
        value="10"
      >
        10
      </option>
    </select>
    <div
      className="page1-button-normal page1-flex page1-right-button"
      onClick={[Function]}
    >
      Next
    </div>
  </div>
</div>
`;
